torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp

```diff
@@ -396,37 +407,116 @@ class DoubleBufferInserter : private kir::ExprMutator {
 
   void insert(
       kir::ForLoop* double_buffer_loop,
-      const std::vector<UnaryOp*>& loads) {
+      const std::vector<Expr*>& loads) {
     auto prologue_loop = DoubleBufferLoopCloner::clone(
         double_buffer_loop, loads, LoopType::Prologue);
     registerInsertBefore(double_buffer_loop, prologue_loop);
 
     auto write_to_smem =
-        std::any_of(loads.begin(), loads.end(), [](const UnaryOp* uop) {
-          return uop->out()->as<TensorView>()->getMemoryType() ==
+        std::any_of(loads.begin(), loads.end(), [](const Expr* expr) {
+          return expr->output(0)->as<TensorView>()->getMemoryType() ==
               MemoryType::Shared;
         });
 
     // RAW sync is not inserted for double buffered tensors. The only
     // exception is the prologue load.
+    bool insert_cpasync_wait = false;
     if (write_to_smem) {
       // Here the initial sync before entering double buffer loop is
-      //  inserted. It will also need to sync with async gmem loads.
+      //  inserted.
+
+      // If any of the double buffered tensor in this double buffer
+      //  loop is async copy. We want to wait for the gmem loads to
+      //  finish before synchronizing the block.
+      if (std::any_of(loads.begin(), loads.end(), ir_utils::isCpAsyncOp)) {
+        auto cp_async_wait = IrBuilder::create<kir::CpAsyncWait>();
+        registerInsertBefore(double_buffer_loop, cp_async_wait);
+        insert_cpasync_wait = true;
+      }
+
+      // Insert the initial block sync before entering main loop.
+      //
+      // TODO: we may not need this sync at all if we didn't need
+      //  to sync the block before. This information is in sync info.
+      //  In this case shared mem double buffering "collapses" into
```

Comment:
Not sure what collapsing means.

Does this mean we don't need block sync when the async wait inserted above is sufficient?

Reply:
Yes, I will shorten the comment or even just do it in this PR. It's basically that double buffer sync should now look at `sync info` to decide if a sync should be inserted. 

It's generic double buffer change but I guess if it's a few lines I could just do it in this PR as well.

Reply:
Sounds good.

Reply:
This is the only part I still need to review. Let me know if you're done.

Reply:
I think I added all I wanted to include about double buffering in this PR. 

Will have a follow up on double buffering for sure, to clear all the todo's.

Reply:
@naoyam This PR should be ready for re-review. 

Will need to merge back other changes before merging this one, but I expect all remaining commits fully mechanical.